File: python_solutions/1469.find-all-the-lonely-nodes.cpp
// Containers
#include <deque>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <vector>
#include <iostream>
#include <numeric>


#include "conf.d/say.h"

using namespace std; 
int dirs[5] = {-1, 0, 1, 0, -1};

// --------------------------------------------------------
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    std::vector<int> res;
    void dfs(TreeNode* r) {
      if (r == nullptr)return ;
      else if (r->left == nullptr && r->right == nullptr) return ;
      else if (r->left == nullptr) { res.push_back(r->right->val); dfs(r->right); }
      else if (r->right == nullptr) { res.push_back(r->left->val); dfs(r->left); }
      else {dfs(r->left); dfs(r->right);}
    }
    vector<int> getLonelyNodes(TreeNode* root) {
      dfs(root);
        return res;
    }
};


int main() {
  Solution s;
  
}

File: python_solutions/1469.find-all-the-lonely-nodes.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007


# https://leetcode.com/problems/find-all-the-lonely-nodes/description/
# Easy
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getLonelyNodes(self, root: TreeNode) -> List[int]:
        def dfs(r):
            if not r or (not r.left and not r.right): return
            if not r.left:
                res.append(r.right.val)
                dfs(r.right)
            elif not r.right:
                res.append(r.left.val)
                dfs(r.left)
            else:
                dfs(r.left)
                dfs(r.right)

        res = []
        dfs(root)
        return res
